plugins/transport-nio/src/main/java/org/elasticsearch/http/nio/NioHttpRequest.java

```diff
@@ -40,24 +42,35 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.stream.Collectors;
 
 public class NioHttpRequest implements HttpRequest {
 
     private final FullHttpRequest request;
-    private final BytesReference content;
     private final HttpHeadersMap headers;
     private final int sequence;
+    private final AtomicBoolean released;
+    private final boolean pooled;
+    private BytesReference content;
 
     NioHttpRequest(FullHttpRequest request, int sequence) {
+        this(request, sequence, new AtomicBoolean(false));
+    }
+
+    private NioHttpRequest(FullHttpRequest request, int sequence, AtomicBoolean released) {
+        this(request, new HttpHeadersMap(request.headers()), sequence, released, request.content() instanceof UnpooledHeapByteBuf == false,
```

Comment:
Technically we could do more involved detection of whether or not the buffer here really is unpooled like was done in https://github.com/elastic/elasticsearch/pull/32228 which might be nice for low-heap nodes that will often see unpooled composite buffers. I just added this check to avoid wasting memory in obvious spots now, but IMO we could optimize with the deeper check for unpooled composite buffers in a follow up.

Reply:
I don't think this is reliable. In Nio these might often be `UnpooledHeapByteBuf` even if they are pooled from BigArrays. It's how we interface with big arrays (`PagedByteBuf`). I think in Nio for right now we should just always do the copy. As in remove the "pooled" optimization.

Reply:
Makes sense, reverted the NIO implementation for now.